### PR TITLE
Update LLM Agents Pipeline README.md

### LINKED ISSUE #1307: Create Sherlock example for LLM agents pipeline
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

High

### Please provide a clear description of problem this feature solves

As part of the Sherlock work, an example showing how to use Morpheus to execute multiple LLM agents which are able to use tools and execute multiple LLM queries in order to solve a complex task.

## Purpose

The purpose of this example is to illustrate how a user could build a pipeline that will integrate an LLM agent into a Morpheus pipeline. This example builds on the previous two examples, #1298 and #1306, by adding the ability to execute multiple LLM queries as part of a chain and also to augment those queries with external information if needed.

Using LLM agents allows the LLMs to be able to execute "tools" that can help them solve their task. For example, if your question was "What is Leonardo DiCaprio's girlfriends age raised to the 0.43 power?", the LLM might not know who Leonardo DiCaprio's girlfriend is, but it will know it can find that out from a quick search. If you perform the internet search for the LLM and ask it the question again with the result from the internet search, it will have the information necessary to properly respond.

## Scenario

This example will show two different implementations of an agent pipeline but the pipeline and components could be used in many scenarios with different requirements. At a high level, the following illustrates different customization points for this pipeline and the specific choices made for this example:

- LLM Service
   - This pipeline could support any type of LLM service which is compatible with our `LLMService` interface.
   - Such services include OpenAI, NeMo, or even running locally using `llama-cpp-python`
   - For this example, we will focus on using OpenAI as the LLM service. This service has been chosen to be different from the previous example which uses NeMo. Additionally, the models in OpenAI have been trained to give responses that work well in the ReAct agent architecture.
- Agent type
   - There are many different styles of agents that are available today. Each one of these agent types alters the pattern for asking the LLM what the next task is, running that operation, and then asking the LLM what to do next. That repeats until the LLM is satisfied with the output.
   - For this example, we will be using the ReAct agent type. This type of agent is the most popular and has proven to work well and be reliable.
- Agent tools
   - Depending on the problem trying to be solved, there are many different tools that can be made available to the LLM agent such as an internet search, VDB retrievers, calculators, Wikipedia, etc.
   - For this example, we will be using the internet search tool and an llm-math tool. This allows the LLM agent to execute arbitrary Google searches and to perform math equations if necessary
- LLM Library
   - Most of the popular LLM libraries, Langchain, llama-index, and Haystack, have the ability to run LLM agents to execute these higher level tasks.
   - For this example, we will be using the Langchain library to provide all of our LLM agents. This allows us to run agents directly in a Morpheus pipeline reducing the overhead of migrating an existing system to Morpheus, and also eliminates the need to replicate the work already done by these libraries.

## Implementation

This example will be composed of two different `click` commands.

### Simple Morpheus agent pipeline

The simple Morpheus pipeline is built using the following components:

1. An `InMemorySourceStage` to hold the LLM queries in a DataFrame
1. A `DeserializationStage` to convert the `MessageMeta` objects into `ControlMessages` needed by the `LLMEngine`
   1. New functionality was added to the `DeserializeStage` to support `ControlMessages` and add a default task to each message.
1. A `LLMEngineStage` then wraps the core `LLMEngine` functionality
   1. An `ExtracterNode` pulls the questions out of the DataFrame
   1. A `LangChainAgentNode` runs the Langchain agent executor for all provided input. This node will utilize the agents `arun` interface to run the agents asynchronously.
   1. Finally, the responses are put back into the `ControlMessage` using a `SimpleTaskHandler`
1. The pipeline concludes with an `InMemorySink` stage to store the results.

### Kafka Morpheus pipeline

The Kafka Morpheus agent pipeline is functionally similar to the simple agent pipeline, however it uses a Kafka message bus to send messages to the pipeline instead of an in memory source. The reason we are using a Kafka message bus is to show the capabilities of the async `LLMEngine` functions running inside of a Morpheus pipeline. Normally, when running Langchain agents using the `arun` interface, they will all run simultaneously. But the Langchain system cannot handle new requests that come in while the current batch is running. In Morpheus, when new messages are sent to the pipeline while existing ones are processing, they will immediately begin processing increasing the parallelism. This helps show why running these LLM agents in a pipeline is advantageous over a normal python library.

The major differences between the simple Morpheus agent pipeline and the Kafka Morpheus agent pipeline are:

- The source for the pipeline is a `KafkaSourceStage` to pull messages off of the bus as they are created
- Message responses are either printed directly to the console or output to another Kakfa topic.

## Completion Criteria

The following items need to be satisfied to consider this issue complete:

- [x] A `README.md` containing information on the following:
   - [x] Background information about the problem at hand
   - [x] Information about the specific implementation and reasoning behind design decisions (use content from this issue)
   - [x] Step-by-step instructions for running the following:
      - [x] How to run the Simple Morpheus agent pipeline
         - [x] Including Information on how to get an OpenAI API key and what environment variable to set.
         - [x] Including information on how to get a Serpapi.com API key and what environment variable to set
         - [x] The command to run the Morpheus pipeline
      - [x] How to run the Kafka Morpheus agent pipeline
         - [x] Including instructions on how to set up the Kafka cluster for sending messages
         - [x] Including Information on how to get an OpenAI API key and what environment variable to set.
         - [x] Including information on how to get a Serpapi.com API key and what environment variable to set
         - [x] Including instructions on how to view the output Kafka topic where responses will be placed
         - [x] The command to run the Morpheus pipeline
   - [x] The README.md should be linked in the developer docs
- [x] A functioning Simple Morpheus agent pipeline command which satisfies the following:
   - [x] Should run without error using all default arguments
   - [x] Correctly run the LLM agents using the OpenAI chat model
   - [x] Provide information about the success or failure of the pipeline. Including the number of queries run, throughput and total runtime.
- [x] A functioning Kafka Morpheus agent pipeline command which satisfies the following:
   - [x] Should run without error using all default arguments
   - [x] Correctly run the LLM agents using the OpenAI chat model
   - [x] Output the responses to the console or to a separate output Kafka topic
   - [x] Provide information about the success or failure of the pipeline. Including the number of queries run, throughput and total runtime.
- [x] Tests should be added which include the following
   - [x] Test successfully running the Simple Morpheus agent pipeline
   - [x] Test successfully running the Kafka Morpheus agent pipeline

### Dependent Issues

The following issues should be resolved before this can be completed:

```[tasklist]
- [ ] #1178 
- [ ] #1179 
- [ ] #1277
- [ ] #1278 
- [ ] #1299
- [ ] #1302 
- [ ] #1303 
```

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by bsuryadevara**
## Description

Updated agents pipeline README.md

Closes #1307 

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

File: examples/llm/agents/README.md
Comment (drobison00) [inline examples/llm/agents/README.md:1]: @bsuryadevara Could you bring the format of this in line a bit more closely with the other 3 README.mds.

See: https://github.com/nv-morpheus/Morpheus/blob/fea-sherlock/examples/llm/vdb_upload/README.md

**Review by drobison00 — CHANGES_REQUESTED** [review]

File: examples/llm/agents/README.md
Comment (bsuryadevara) [reply examples/llm/agents/README.md:1]: Sure

Comment (bsuryadevara) [reply examples/llm/agents/README.md:1]: @drobison00 Updated as suggested.

**Review by drobison00 — APPROVED** [review]

**Pull request merged**